distribution/lib/Standard/Test/0.0.0-dev/src/Main.enso

```diff
@@ -196,9 +196,43 @@ Any.should self verb argument = verb Verbs self argument
          example_fail = Test.fail "Something went wrong."
 fail : Text -> Assertion
 fail message details=Nothing =
-    failure = Failure message details
+    clue_data = State.get Clue
+    message_with_clue = case clue_data of
+        Clue_Data add_clue -> add_clue message
+        _                  -> message
+    failure = Failure message_with_clue details
     Panic.throw failure
 
+## Executes the block of code passed as behavior and adds a clue modifier which
+   changes how assertion failures are reported.
+
+   Nesting with_clue invocations results in clue aggregation.
+
+   Arguments:
+   - clue: either a text which gets prepended to the failure or a function which transforms the failure message
+   - behavior: the behavior to test
+
+   > Example
+     Add a clue to a test
+        import Standard.Test
+        main = Test.Suite.run_main <|
+            Test.group "Tests" <|
+                Test.specify "some property" <|
+                    xs = Vector.new 100 (n -> n)
+                    xs.each x->
+                        Test.with_clue ("[" x.to_text "] ") <|
+                                x . should_equal 0
+
+with_clue : Text|(Text -> Text) -> Any -> Any
+with_clue ~clue ~behavior =
+    add_clue = if clue.is_a Text then (x -> clue + x) else clue
+    prev_clue = State.get Clue
+    next_clue = case prev_clue of
```

Comment:
Shouldn't there be a test that shows this _"composability of clues"_ really works?

Reply:
Well yes, but how?
Pragmatically, writing a test with nested with_clues would be ok, but..
(TL;DR: I would suggest somebody manually verifies this for now as an automated test is ultimately unreliable)

`with_clue` is not strictly an adapter or somesuch - to make it work, i made (small) changes to the way tests are run.

So, in a way, running a test on `with_clue` runs part of the code under test as part of the test framework.
Which is not helpful, since it requires the assumption that the code under test works as intended.
(Pathogenic example: the changes of this commit would introduce behavior such that all tests pass; 
worse even: the changes of this commit would introduce behavior such that all tests on with_clue pass)

Which brings up the bigger question: Is it possible to properly test the test framework?

I do not think so, unless one defers to another testing framework.
So, to verify properties about the code, one would have to formally prove them, which would be much easier by using types once they are available.

Pragmatically though, I would suggest that somebody should manually test any desirable properties of the test framework.
Once the test framework is considered stable and unchanging, somebody shall go about to formally prove any important properties (with or without the help of a type checker).

Another viable route is to use an existing testing framework which is reasonably stable to test enso's test framework (of which `with_clue` is a part).

Reply:
> Is it possible to properly automate testing the test framework?

Of course, it is! Check [JUnit4 tests](https://github.com/junit-team/junit4/tree/main/src/test/java) written in JUnit4.

Reply:
@diab0l did you manage to check out the tests that @JaroslavTulach mentioned above?

Reply:
@xvcgreg Yes, i did.
From what I can tell, the JUnit4 tests make reasonable assumptions about the runner used to run the tests.
There are tests on the runner, but they require a reasonable runner to run them.

The other tests defer, where possible, to not using the test framework, but instead catching and throwing exceptions manually, etc.

Frankly, I am not sure how this translates to enso.
Last time I checked, there were no tests on the testing framework.

Should I go ahead and create the first one?

Reply:
That'd be awesome :)
@JaroslavTulach